### PR TITLE
deprecate Surface_mesher package

### LINKED ISSUE #3237: Future of Surface_mesher
AFAIK, the only place where Surface_mesher is hard-coded is in Poisson surface reconstruction.

What is the plan for the future of `Surface_mesher`?
- remove the package and recommend using Mesh_3
- offer full compatibility: it means having wrapper for oracle + c2t3 working with Mesh_3
- ?

I open the issue to track the topic. The second option looks better from a user point of view.

Related issues:
- license checking of Surface_mesher should be using Mesh_3 in the end
- the macro `BLABLA_DEMO_USING_SURFACE_MESHER` in the demo can be removed
- ?

@lrineau @maxGimeno @MaelRL @janetournois please complete the issue if you see some important points to consider.

### LINKED ISSUE #8211: Dependencies file with tag files and the resulting doxygen "dxy"  file
With the fix b9ee8c6 a fix (by @sloriot) was made to fix the issue:
```
error: Tag file '.../doc_tags/Lab.tag' does not exist or is not a file. Skipping it...
```
when building the documentation of the (not really supported) Three  package.
The fix as is is correct and worked for a clean build, but when using an existing build it didn't work as the `Three.dxy` configuration file was not rebuild and a remake of the cmake files was not triggered.
The change was made in the file `Three/doc/Tree/dependencies` and this is not listed as a dependency to regenerate the `Three.dxy`

In an email discussion it was pointed out by @lrineau that:
> The answer is probably https://cmake.org/cmake/help/latest/prop_dir/CMAKE_CONFIGURE_DEPENDS.html

The solution lies probably in Documentation/doc/CMakeLists.txt where a configuration dependency between:
`${depend}/doc/${depend}/dependencies` and `${CGAL_DOC_DXY_DIR}/${depend}.dxy` has to be created, where `depend` is a package name).

### PR DESCRIPTION
**Pull request by sloriot**
Surface_mesher has not been updated for a long time and all its functionalities (and more) have been implemented in Mesh_3. We now officially recommend to use Mesh_3 instead by deprecating the package.

Fixes  #3237

Fixes #8211

### CODE REVIEW

**Comment by janetournois** [comment]
There is one issue left in `CGAL/poisson_refine_triangulation.h` , that still uses `Surface_mesh_traits_generator_3`

**Comment by lrineau** [comment]
I am quite surprised by this PR. I thought the urgency for CGAL-6.0-beta1 was to announce a deprecation:

  - in the user-manual,
  - in the `CHANGES.md`,
  - and maybe only for the entry-point function `make_surface_mesh`.
  
This PR is deprecating all the headers, and then everything that uses Surface_mesher internal now has to move in a rush, before CGAL-6.0-beta1. Is there a chance that:

  - we can converge to a stable state in, say, less than a week,
  - and that we have the confidence we will not introduce regression bugs in the process?

**Comment by lrineau** [comment]
With commit 6547f648968441b60f7bc4733bda687b69ad3fa5 I have repaired the reconstruction plugin, and tested it in CGAL Lab. It seems correct.

Screenshot_20240605_165157

On the console:

```
Computes Poisson implicit function using Conjugate Gradient...
Construction time of the sizing field: 1.2761150000000008 seconds
Total implicit function (triangulation+refinement+solver): 5.9814699999999998 seconds
Surface meshing...
/home/lrineau/Git/cgal-master/Poisson_surface_reconstruction_3/include/CGAL/Poisson_reconstruction_function.h:263:70: runtime error: reference binding to null pointer of type 'struct value_type'
Surface meshing: 10.593293999999998 seconds, 756 output vertices
Total reconstruction (implicit function + meshing): 16.730679000000002 seconds
Reconstruction error:
  max = 0.127707182737539
  avg = 0.0087130995426717341
Reconstruction achieved in 10.024125099182129s
```

Note that the UBAN sanitizer does not like that code, for a null cell pointer:

https://github.com/CGAL/cgal/blob/6547f648968441b60f7bc4733bda687b69ad3fa5/Poisson_surface_reconstruction_3/include/CGAL/Poisson_reconstruction_function.h#L261-L264

**Comment by lrineau** [comment]
> Note that the UBAN sanitizer does not like that code, for a null cell pointer:
> 
> https://github.com/CGAL/cgal/blob/6547f648968441b60f7bc4733bda687b69ad3fa5/Poisson_surface_reconstruction_3/include/CGAL/Poisson_reconstruction_function.h#L261-L264

Fixed by 9cd4535b11877266e8085be3299ea81df4ed5123. Now:

https://github.com/CGAL/cgal/blob/9cd4535b11877266e8085be3299ea81df4ed5123/Poisson_surface_reconstruction_3/include/CGAL/Poisson_reconstruction_function.h#L261-L267

Maybe `s_iterator_to` in the compact container should have an overload for pointers, to avoid that extract test for null.

File: Poisson_surface_reconstruction_3/include/CGAL/Surface_mesher/Poisson_sphere_oracle_3.h [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+//
+// Author(s)     : Laurent RINEAU
+
+#ifndef CGAL_POISSON_SURFACE_MESHER_POISSON_SPHERE_ORACLE_3_H
+#define CGAL_POISSON_SURFACE_MESHER_POISSON_SPHERE_ORACLE_3_H
+
+#include <CGAL/license/Surface_mesher.h>
```
Comment (lrineau) [inline Poisson_surface_reconstruction_3/include/CGAL/Surface_mesher/Poisson_sphere_oracle_3.h:16]: TODO: fix copyright notice (date).

TODO: discussion about headers names (including path)

Plus the license-check header:
```suggestion
// Copyright (c) 2006-2007  INRIA Sophia-Antipolis (France).
// All rights reserved.
//
// This file is part of CGAL (www.cgal.org).
//
// $URL$
// $Id$
// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
//
//
// Author(s)     : Laurent RINEAU

#ifndef CGAL_POISSON_SURFACE_MESHER_POISSON_SPHERE_ORACLE_3_H
#define CGAL_POISSON_SURFACE_MESHER_POISSON_SPHERE_ORACLE_3_H

#include <CGAL/license/Poisson_surface_reconstruction_3.h>
```

File: Surface_mesher/doc/Surface_mesher/dependencies [hunk 2]
```diff
@@ -6,4 +6,5 @@ Circulator
 Stream_support
 Triangulation_2
 Triangulation_3
+Mesh_3
```
Comment (lrineau) [inline Surface_mesher/doc/Surface_mesher/dependencies:9]: That is surprising.

TODO: check why

File: Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction_example.cpp [hunk 3]
```diff
@@ -1,3 +1,6 @@
+/// @TODO: change the usage from Surface_mesher to Mesh_3
+#include <CGAL/Installation/internal/disable_deprecation_warnings_and_errors.h>
+
```
Comment (lrineau) [inline Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction_example.cpp:3]: It think we (I?) should update this example.

File: Poisson_surface_reconstruction_3/include/CGAL/Mesh_3/Poisson_mesh_traits_generator_3.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Laurent Rineau, Jane Tournois
+
+#ifndef CGAL_MESH_3_POISSON_MESH_TRAITS_GENERATOR_3_H
+#define CGAL_MESH_3_POISSON_MESH_TRAITS_GENERATOR_3_H
```
Comment (lrineau) [inline Poisson_surface_reconstruction_3/include/CGAL/Mesh_3/Poisson_mesh_traits_generator_3.h:13]: This header file should probably be renamed/moved.

File: Poisson_surface_reconstruction_3/test/Poisson_surface_reconstruction_3/poisson_reconstruction_test.cpp [hunk 5]
```diff
@@ -8,6 +8,10 @@
 // poisson_reconstruction_test mesh1.off point_set2.xyz...
 
 // CGAL
+
+/// @TODO: change the usage from Surface_mesher to Mesh_3
+#include <CGAL/Installation/internal/disable_deprecation_warnings_and_errors.h>
+
 #include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
```
Comment (lrineau) [inline Poisson_surface_reconstruction_3/test/Poisson_surface_reconstruction_3/poisson_reconstruction_test.cpp:15]: Maybe we should keep this test, to test the backward compatibility.

But then it should be duplicated with another one using Mesh_3 instead.

**Comment by lrineau** [comment]
> There was an error while building the doc:
> 
> ```
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor classcgal_multimap found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor sectionProjectionFunctionObjects found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor Chapter_STL_Extensions_for_CGAL found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_intro found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_doubly found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_compact found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_multi found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_hash found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_polyobject found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_uncertainty found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_complexity found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_defaults found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor secchecks found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_alteriung found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_control found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_customizing found
> /home/runner/work/cgal/cgal/build_doc/doc_tags/STL_Extension.tag:5046: warning: Duplicate anchor stl_failure_example found
> ```
> 
> https://github.com/CGAL/cgal/actions/runs/9388765120

@sloriot Can you help? Or maybe @albert-github?

File: Surface_mesher/doc/Surface_mesher/dependencies (hunk 2 above)
Comment (MaelRL) [reply Surface_mesher/doc/Surface_mesher/dependencies:9]: Note that this is doc/dependencies, not the include's dependencies.

Comment (lrineau) [reply Surface_mesher/doc/Surface_mesher/dependencies:9]: Oh, right! Let's cl;ose this conversation.

**Comment by albert-github** [comment]
It looks like that the `STL_Extension.tag` is mentioned twice in the used Doxyfile (or the content is contained in another tag file as well).
It looks like that this is a doxygen 1.9.6 problem and is fixed in the doxygen 1.10.0 and later, looks like fixes https://github.com/doxygen/doxygen/pull/10323 it looks though that also 1.11.0 version contains a small adjustment fix https://github.com/doxygen/doxygen/issues/10850 i.e. https://github.com/doxygen/doxygen/pull/10854

In the 1.11.0 version it looks like the message cannot appear anymore except for the C++-20 modules

File: Surface_mesher/doc/Surface_mesher/dependencies [hunk 6]
```diff
@@ -6,4 +6,6 @@ Circulator
...
 Polyhedron
+STL_Extension
```
Comment (albert-github) [inline Surface_mesher/doc/Surface_mesher/dependencies:11]: Was already present (line 3)

**Review by albert-github — CHANGES_REQUESTED** [review]

**Comment by lrineau** [comment]
> The documentation is built. It will be available, after a few minutes, here: https://cgal.github.io/8248/wip/Manual/index.html

Nice! Now the doc builds without any errors.

> It looks like that the `STL_Extension.tag` is mentioned twice in the used Doxyfile (or the content is contained in another tag file as well).

Thanks @albert-github. That was indeed the issue. I have sorted the two `doc/*/dependencies` files of this PR, so ease future duplication errors. See https://github.com/CGAL/cgal/pull/8248/commits/36c7ecb77375d03014f5571858c02d045a527381 and https://github.com/CGAL/cgal/pull/8248/commits/6c6814dd2fd876a330a7a1b855767dcd70a2b248.

**Comment by lrineau** [comment]
## Issue with `Poisson_implicit_surface_3`

For the moment, in `master`, we have an improved point oracle for Surface_mesher, than is only used one example file: `Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction.cpp`.

### History

_Thanks to `git log -p --reverse -S Poisson_implicit_surface_3 cgal/master`._

- The class template `CGAL::Poisson_implicit_surface_3` (and its companion `Poisson_implicit_surface_oracle_3`) was introduced by commit b0ce738c82988226af8cdf7ce2f029a3a30e9574 (Sep 27, 2011, first merged in CGAL by commit 485d1bc5e0202ac54a512216473a2674be344962, Nov 20, 2012, for CGAL-4.2). In the same branch `Surface_reconstruction_points_3-make_it_faster-afabri`, there was also the following commits:
  - 197870acffe4359940c107f31676fe5885007cff adding the use of that better oracle class to the poisson reconstruction demo (`demo/Surface_reconstruction_points_3/PS_demo_poisson_plugin_cgal_code.cpp`)
  - eb1dc0775047fbed03d9f731c1e3a1b33755b3f5 adding the use of the improved oracle to the examples `examples/Surface_reconstruction_points_3/poisson_reconstruction.cpp`.
- then 028b6cc505a4fe5e4fc0f226f21af275566d6793 (in CGAL-4.5):
  > remove the Point set demo as all functionalities are now in the Polyhedron demo
  
  but the polyhedron demo never had the speed improvements
- 2016, @sgiraudot added the improved oracle in `examples/Poisson_surface_reconstruction_3/tutorial_example.cpp`, commit f052a95d2ab3a0bd6b15e28aed8d6c2721b6c369, but then removed that usage in b9730aa925cdaaeada768c129afd2afb63177ead in 2020.

**Comment by lrineau** [comment]
> ## Issue with `Poisson_implicit_surface_3`
> 
> For the moment, in `master`, we have an improved point oracle for Surface_mesher, than is only used one example file: `Poisson_surface_reconstruction_3/examples/Poisson_surface_reconstruction_3/poisson_reconstruction.cpp`.

And that example is not in the documentation.

**Comment by lrineau** [comment]
> Issue with `Poisson_implicit_surface_3`

The difference is in the associated oracle class. The Poisson function is queried via a hidden dedicated API:

https://github.com/CGAL/cgal/blob/2c13f1fe6bcb7496b5f5c68095790ae4f894f3dc/Poisson_surface_reconstruction_3/include/CGAL/Surface_mesher/Poisson_implicit_surface_oracle_3.h#L286-L292

And then, later in the bisection, those added lines:

https://github.com/CGAL/cgal/blob/2c13f1fe6bcb7496b5f5c68095790ae4f894f3dc/Poisson_surface_reconstruction_3/include/CGAL/Surface_mesher/Poisson_implicit_surface_oracle_3.h#L303-L310

Once the two endpoints of the segment are in the same cell, then there is no need to continue the bisection.

**Comment by lrineau** [comment]
> Issue with `Poisson_implicit_surface_3`

The full diff is at https://gist.github.com/lrineau/f5408f2e0c57e35ebf13f48b4778db90

**Comment by lrineau** [comment]
@sloriot @afabri See 6daec19ce91987970959af9bd48042a8bf49c659. There are a lot more dependencies for Poisson reconstruction, now.

**Review by lrineau — APPROVED** [review]
Looks good, now.

- Tested a few packages with CTest: ✔️
- Tested CGAL Lab with the reconstruction and xyz plugins: ✔️

```
100% tests passed, 0 tests failed out of 138

Label Time Summary:
Basic_viewer_Examples                        =  71.67 sec*proc (14 tests)
CGAL_build_system                            =   2.86 sec*proc (2 tests)
Installation                                 =   2.86 sec*proc (2 tests)
Poisson_surface_reconstruction_3_Examples    = 129.51 sec*proc (10 tests)
Poisson_surface_reconstruction_3_Tests       = 235.40 sec*proc (10 tests)
STL_Extension_Examples                       =  29.91 sec*proc (14 tests)
STL_Extension_Tests                          = 280.70 sec*proc (72 tests)
Surface_mesher_Examples                      =  32.04 sec*proc (6 tests)
Surface_mesher_Tests                         =  60.02 sec*proc (10 tests)

Total Test time (real) = 152.82 sec
```

**Comment by lrineau** [comment]
In CGAL Lab, the dialog that opens with the action "Surface Reconstruction", still mention "Surface Mesher". But I do not know what it could be replaced with. "Mesh_3" and "3D Mesh Generation" both the same issue: they do not convey that the code is only returning a surface mesh, as a result.

Screenshot_20240607_162133 (1)

**Comment by afabri** [comment]
This is no problem. We use Mesh_3 as Surface Mesher.  In fact even with Mesh_3 we somehow would lilke to have a function called make_surface_mesh.

**Comment by sloriot** [comment]
Successfully tested in CGAL-6.0-Ic-264

**Pull request merged**